rasa/core/processor.py

```diff
@@ -374,8 +376,33 @@ def predict_next_action(
         """Predicts the next action the bot should take after seeing x.
 
         This should be overwritten by more advanced policies to use
-        ML to predict the action. Returns the index of the next action.
+        ML to predict the action.
+
+        Returns:
+             The index of the next action and prediction of the policy.
+
+        Raises:
+            ActionLimitReached if the limit of actions to predict has been reached.
         """
+        should_predict_another_action = self.should_predict_another_action(
+            tracker.latest_action_name
+        )
+        reversed_events = list(tracker.events)[::-1]
+        num_predicted_actions = 0
+
+        for e in reversed_events:
+            if isinstance(e, ActionExecuted):
+                if e.action_name in (ACTION_LISTEN_NAME, ACTION_SESSION_START_NAME):
+                    break
+                num_predicted_actions += 1
+
+        if self.is_action_limit_reached(
+            num_predicted_actions, should_predict_another_action
+        ):
+            raise ActionLimitReached(
```

Comment:
Don't we need to handle this in `_predict_and_execute_next_action` now?

Reply:
This [implementation proposal section](https://www.notion.so/rasa/rasa-core-test-192bb2cb16e6430ba08c5bf53c78d6d2) indicates to adapt `processor.predict_next_action` - I think mainly because `_run_action_prediction` in `rasa.core.test` module uses this method and not `processor._predict_and_execute_next_action`.

Reply:
Sure,  but `_predict_and_execute_next_action` calls `predict_next_action` so we need to make sure it still works correctly after this change.

Reply:
Also this ticket predates that implementation proposal right? It was created because of the revamp so we can make the interface to `GraphProcessor` smaller. (Although it of course overlaps)

Reply:
How would you test that `_predict_and_execute_next_action` works correctly after this change, apart from using CI checks?

Reply:
> this ticket predates that implementation proposal right?

The ticket doesn't predate the implementation proposal -  GH indicates that the ticket was created on Jul 15, that was last week Thu after the implementation proposal was submitted for review.
I also checked in advance the usages of `_predict_and_execute_next_action` and this is not used anywhere in `rasa.core.test`.

Reply:
Sorry that was a miss-understanding - I was referring to the Test Stories implementation proposal which also lists this as a change! (But I see that you were actually linking to the Architecture revamp 👍 )